torch/distributed/rpc/rref_proxy.py

```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
 def _local_invoke_async_execution(rref, func_name, args, kwargs):
     return getattr(rref.local_value(), func_name)(*args, **kwargs)
 
+# convert a Future[Future[X]] to a Future[X]
+def _unwrap_future(fut: Future) -> Future:
+    res: Future = Future()
+
+    def _unwrap_level0(fut0):
+        try:
+            res.set_result(fut0.wait())
+        except BaseException as ex:
+            res.set_exception(ex)
+
+    def _unwrap_level1(fut1):
+        try:
+            fut1.wait().then(_unwrap_level0)
+        except BaseException as ex:
+            res.set_exception(ex)
+
+    fut.then(_unwrap_level1)
+    return res
+
 def _invoke_rpc(rref, rpc_api, func_name, timeout, *args, **kwargs):
-    # Since rref._get_type can potentially issue an RPC, it should respect the
-    # passed in timeout here.
-    rref_type = rref._get_type(timeout=timeout)
-
-    _invoke_func = _local_invoke
-    # Bypass ScriptModules when checking for async function attribute.
-    bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
-        rref_type, torch._C.ScriptModule
-    )
-    if not bypass_type:
-        func = getattr(rref_type, func_name)
-        if hasattr(func, "_wrapped_async_rpc_function"):
-            _invoke_func = _local_invoke_async_execution
-
-    return rpc_api(
-        rref.owner(),
-        _invoke_func,
-        args=(rref, func_name, args, kwargs),
-        timeout=timeout
-    )
+    # this function returns either the value or a Future depending on what rpc_api is.
+    def _rref_type_cont(future, is_future):
+        if is_future:
+            rref_type = future.wait()
```

Comment:
Don't we need to chain callbacks here instead of just calling `wait()` on the future? Also, isn't `is_future` always going to be `False` in this code?

Reply:
This callback is the one chained from the first future, we need to call `wait()`.
Calling `wait()` from the callback to extract the value is the suggested way to do so.
`is_future` is false when following the blocking path.

Reply:
Ah I see, basically `future.wait()` won't wait at all. In that case, I think just for readability sake, it would be better to use `future.value()` since the invariant here is that the future has completed. Now if this invariant is not true for some reason (bugs) this would throw an error instead of blocking a thread with `future.wait()`.

Reply:
Makes sense, I'll update the docs for `then` separaredly.